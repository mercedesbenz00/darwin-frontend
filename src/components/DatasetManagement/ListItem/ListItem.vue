<template>
  <div
    class="list-item"
    :class="{'list-item--selected': isSelected}"
    @click.stop.prevent="onSelect"
  >
    <div class="list-item__checkbox">
      <check-box
        v-if="!isFolder"
        name="item-selected"
        :value="isSelected"
        size="small"
      />
    </div>
    <div class="list-item__content">
      <div class="list-item__thumbnail">
        <folder-icon
          v-if="isFolder"
          class="list-item__thumbnail__folder"
        />
        <img
          v-else
          v-lazy="thumbnail"
          class="list-item__thumbnail__img"
        >
        <video-overlay-icon
          v-if="isVideo"
          class="list-item__thumbnail__video_overlay"
        />
        <frames-overlay-icon
          v-if="isVideoFrames"
          class="list-item__thumbnail__video_overlay"
        />

        <div
          v-if="$slots['overlay']"
          class="list-item__overlay"
        >
          <slot name="overlay" />
        </div>
      </div>
      <div class="list-item__file_name">
        <dataset-item-name :name="name" />
      </div>
      <div class="list-item__status">
        <slot name="status" />
      </div>
      <div class="list-item__created_at">
        {{ formattedDate }}
      </div>
      <slot name="details" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import { FolderIcon } from '@/assets/icons/V1'
import CheckBox from '@/components/Common/CheckBox/V1/CheckBox.vue'
import DatasetItemName from '@/components/DatasetManagement/Common/DatasetItemName.vue'
import FramesOverlayIcon from '@/components/DatasetManagement/assets/frames-listitem-overlay.svg?inline'
import VideoOverlayIcon from '@/components/DatasetManagement/assets/video-listitem-overlay.svg?inline'
import { formatDate } from '@/utils'

/**
 * Defines layout and look for all list item components.
 *
 * Renders date, image and name for all of them.
 * Provides slots for status and details.
 */
@Component({
  name: 'list-item',
  components: { CheckBox, DatasetItemName, FolderIcon, FramesOverlayIcon, VideoOverlayIcon }
})
export default class ListItem extends Vue {
  @Prop({ required: true, type: String })
  name!: string

  @Prop({ type: String, default: null })
  date!: string | null

  @Prop({ type: String, default: null })
  thumbnail!: string

  @Prop({ default: 'image' })
  type!: 'folder' | 'image' | 'video' | 'video-frames'

  @Prop({ default: false, type: Boolean })
  isSelected!: boolean

  get formattedDate (): string {
    return this.date ? formatDate(this.date, 'DD/MM/YY') : ''
  }

  get isFolder () {
    return this.type === 'folder'
  }

  get isVideo () {
    return this.type === 'video'
  }

  get isVideoFrames () {
    return this.type === 'video-frames'
  }

  onSelect (evt: MouseEvent) {
    this.$emit('select', evt, !this.isSelected)
  }
}
</script>

<style lang="scss" scoped>
.list-item {
  @include row--center;
  width: 100%;
  height: 35px;
  min-height: 35px;
  background: transparent;
  padding: 2px 0 2px 10px;
  border-radius: $border-radius-default;
  border: 1px solid transparent;
  cursor: pointer;

  &:hover {
    background: $colorAliceShade;
  }
}

.list-item__checkbox {
  @include noSelect;
}

.list-item--selected {
  background: $colorAliceShade;
  border-color: $colorAliceNight;
}

.list-item__content {
  flex: 1;
  @include row--center;
  @include typography(md, default);
  color: $colorSecondaryDark1;
  border-radius: $border-radius-default;
  height: 35px;
  padding-right: 10px;
  overflow: hidden;
}

.list-item__thumbnail {
  @include noSelect;
  @include row--center;
  height: 33px;
  border-radius: $border-radius-default;
  overflow: hidden;
  position: relative;

  &:hover {
    .list-item__overlay {
      display: flex;
    }
  }
}

.list-item__thumbnail__folder {
  width: 20px;
  color: $colorAliceNight;
}

.list-item__thumbnail__img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  overflow: hidden;
}

.list-item__thumbnail__video_overlay {
  position: absolute;
  top: 2px;
  bottom: 52px;
  height: calc(100% - 4px);
}

.list-item__overlay {
  position: absolute;
  @include fullsize;
  overflow: hidden;
  display: none;

  &:hover {
    background: rgba(145, 169, 192, .5);
    @include row--center;
  }
}

.list-item__name {
  @include ellipsis(1, md);
}

.list-item__status {
  @include row;
  align-items: center;
}

.list-item__status > :first-child {
  margin-right: 5px;
}

.list-item__date {
  text-align: right;
}
</style>
