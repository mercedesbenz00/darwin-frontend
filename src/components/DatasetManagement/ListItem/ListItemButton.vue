<template>
  <router-link
    v-if="!!to"
    class="list-item-button"
    :class="disabled && 'list-item-button--disabled'"
    :to="to"
    @click.native.stop
  >
    <slot />
  </router-link>
  <div
    v-else
    class="list-item-button"
    :class="disabled && 'list-item-button--disabled'"
    @click.stop.prevent="$emit('click')"
  >
    <slot />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { Location } from 'vue-router'

/**
 * Renders a button on top of a list item component.
 *
 * The button can be rendered as a router-link, or as a div which emits the
 * click event.
 *
 * The component supports a purely presentational `disabled` prop, in that it
 * will render differently, but it will still be interactive and emit `click.`
 */
@Component({ name: 'list-item-button' })
export default class ListItemButton extends Vue {
  @Prop({ required: false, default: null })
  to!: string | Location | null

  @Prop({ required: false, type: Boolean, default: false })
  disabled!: boolean
}
</script>

<style lang="scss" scoped>
.list-item-button {
  width: 58px;
  height: 28px;
  border: 1px solid $colorWhite;
  border-radius: 3px;
  color: $colorWhite;
  @include row--center;
  @include typography(xs, default, bold);
  text-align: center;
  letter-spacing: 0.5px;
  opacity: 1;
  cursor: pointer;
  z-index: 999;
  text-transform: uppercase;
  transition: all .2s;

  &:hover,
  &:active {
    background: rgba(255, 255, 255, .25);
  }
}

.list-item-button--disabled {
  opacity: 0.7;

  &:hover,
  &:active {
    background: transparent;
  }
}
</style>
