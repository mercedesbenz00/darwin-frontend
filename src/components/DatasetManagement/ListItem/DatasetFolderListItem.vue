<template>
  <list-item
    class="folder-list-item"
    type="folder"
    :name="name"
  >
    <template #overlay>
      <list-item-button :to="folderLocation">
        Open
      </list-item-button>
    </template>
  </list-item>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'

import ListItem from '@/components/DatasetManagement/ListItem/ListItem.vue'
import ListItemButton from '@/components/DatasetManagement/ListItem/ListItemButton.vue'
import GalleryDatasetFolder from '@/components/DatasetManagement/mixins/GalleryDatasetFolder'

/**
 * Used to render folders in dataset management and open datasets.
 *
 * Used when the gallery is in list view mode.
 */
@Component({
  name: 'dataset-folder-list-item',
  components: { ListItem, ListItemButton },
  mixins: [GalleryDatasetFolder]
})
export default class DatasetFolderListItem extends Vue {}
</script>

<style lang="scss" scoped>
@import "ListItemHeader/workflow-list-item";

.list-item {
  @include workflow-list-item;
}

.status-icon-container {
  width: 20px;
  height: 20px;
}
</style>
