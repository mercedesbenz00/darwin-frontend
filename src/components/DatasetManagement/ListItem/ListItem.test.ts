import { createLocalVue, shallowMount } from '@vue/test-utils'
import VueLazyload from 'vue-lazyload'

import ListItem from '@/components/DatasetManagement/ListItem/ListItem.vue'

const localVue = createLocalVue()
localVue.use(VueLazyload)

const stubs = ['check-box']
const date = '2000-01-01T00:00:00'

it('matches snapshot', () => {
  const propsData = { date, name: '0001', thumbnail: 'foo.png' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when given thumbnail', () => {
  const propsData = { date, name: '0001', thumbnail: 'foo.png' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when folder', () => {
  const propsData = { date, name: 'folder', type: 'folder' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when video', () => {
  const propsData = { date, name: '0001', thumbnail: 'foo.png', type: 'video' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when video-frames', () => {
  const propsData = { date, name: '0001', thumbnail: 'foo.png', type: 'video-frames' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when selected', () => {
  const propsData = { date, name: '0001', isSelected: true, thumbnail: 'foo.png' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('emits "select" when clicked', () => {
  const propsData = { date, name: '0001', isSelected: true, thumbnail: 'foo.png' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  wrapper.find('.list-item').trigger('click')
  expect(wrapper.emitted().select).toBeDefined()

  wrapper.setProps({ isSelected: false })

  wrapper.find('.list-item').trigger('click')
  expect(wrapper.emitted().select).toBeDefined()
})

it('emits "select" when checkbox emits "click" ', () => {
  const propsData = { date, name: '0001', isSelected: true, thumbnail: 'foo.png' }
  const wrapper = shallowMount(ListItem, { localVue, propsData, stubs })
  wrapper.find('.list-item').trigger('click')
  expect(wrapper.emitted().select).toBeDefined()

  wrapper.setProps({ isSelected: false })

  wrapper.find('.list-item').trigger('click')
  expect(wrapper.emitted().select).toBeDefined()
})
