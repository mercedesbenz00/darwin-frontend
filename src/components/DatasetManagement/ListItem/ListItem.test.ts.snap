// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="list-item">
  <div class="list-item__checkbox">
    <check-box-stub name="item-selected" type="rect" size="small"></check-box-stub>
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail"><img class="list-item__thumbnail__img">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="0001"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;

exports[`matches snapshot when folder 1`] = `
<div class="list-item">
  <div class="list-item__checkbox">
    <!---->
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail">
      <folder-icon-stub class="list-item__thumbnail__folder"></folder-icon-stub>
      <!---->
      <!---->
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="folder"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;

exports[`matches snapshot when given thumbnail 1`] = `
<div class="list-item">
  <div class="list-item__checkbox">
    <check-box-stub name="item-selected" type="rect" size="small"></check-box-stub>
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail"><img class="list-item__thumbnail__img">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="0001"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;

exports[`matches snapshot when selected 1`] = `
<div class="list-item list-item--selected">
  <div class="list-item__checkbox">
    <check-box-stub name="item-selected" value="true" type="rect" size="small"></check-box-stub>
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail"><img class="list-item__thumbnail__img">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="0001"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;

exports[`matches snapshot when video 1`] = `
<div class="list-item">
  <div class="list-item__checkbox">
    <check-box-stub name="item-selected" type="rect" size="small"></check-box-stub>
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail"><img class="list-item__thumbnail__img">
      <video-overlay-icon-stub class="list-item__thumbnail__video_overlay"></video-overlay-icon-stub>
      <!---->
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="0001"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;

exports[`matches snapshot when video-frames 1`] = `
<div class="list-item">
  <div class="list-item__checkbox">
    <check-box-stub name="item-selected" type="rect" size="small"></check-box-stub>
  </div>
  <div class="list-item__content">
    <div class="list-item__thumbnail"><img class="list-item__thumbnail__img">
      <!---->
      <frames-overlay-icon-stub class="list-item__thumbnail__video_overlay"></frames-overlay-icon-stub>
      <!---->
    </div>
    <div class="list-item__file_name">
      <dataset-item-name-stub name="0001"></dataset-item-name-stub>
    </div>
    <div class="list-item__status"></div>
    <div class="list-item__created_at">
      01/01/00
    </div>
  </div>
</div>
`;
